### PR TITLE
[Feature] Allow next stage to run from scratch

### PR DESCRIPTION
**Pull request by reyzharkov**
## Description

Allows next stage to run from scratch (epoch 0, step 0), which is useful if the training/validation dataset is changing through time (metrics incompatible), e.g. we trained on synthetic data and then finetune on real data.

```yaml
stage1:
  ...
stage2:
  state_params:
    migrate: False  # (default True)
```

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] I have read the [Code of Conduct](https://github.com/catalyst-team/catalyst/blob/master/CODE_OF_CONDUCT.md) document.
- [x] I have read the [Contributing](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md) guide.
- [x] I have read I need to click 'Login as guest' to see Teamcity build logs
- [x] I have checked the code-style using `make check-codestyle`.
- [x] I have written the docstring in Google format for all the methods and classes that I used.
- [x] I have checked the docs using `make check-docs`.

### CODE REVIEW

File: catalyst/core/runner.py [hunk 1]
```diff
@@ -154,7 +154,8 @@ def _get_experiment_components(
     def _prepare_for_stage(self, stage: str):
         utils.set_global_seed(self.experiment.initial_seed)
         migrating_params = {}
-        if self.state is not None:
+        migrate = self.experiment.get_state_params(stage).get("migrate", True)
+        if self.state is not None and migrate:
```
Comment (Scitator) [inline catalyst/core/runner.py]: could you please create something like `stage_state_params`, and reuse them during State creation? I just don't like the idea for several `self.experiment.get_state_params` call, if there is no need in it

**Review by Scitator — CHANGES_REQUESTED** [review]

File: catalyst/core/runner.py [hunk 2]
```diff
@@ -154,7 +154,8 @@ def _get_experiment_components(
     def _prepare_for_stage(self, stage: str):
         utils.set_global_seed(self.experiment.initial_seed)
         migrating_params = {}
-        migrate = self.experiment.get_state_params(stage).get("migrate", True)
+        stage_state_params = self.experiment.get_state_params(stage)
+        migrate = stage_state_params.get("migrate", True)
```
Comment (Scitator) [inline catalyst/core/runner.py]: what do you think about `.pop` usage?

Comment (reyzharkov) [reply catalyst/core/runner.py]: maybe `migrate` will be important for some callbacks, so let it be?

Comment (Scitator) [reply catalyst/core/runner.py]: 🤔 what about naming then? maybe `migrate_stage=[True/False]`?

Comment (reyzharkov) [reply catalyst/core/runner.py]: You mean to rename `migrate` -> `migrate_stage`?

Comment (Scitator) [reply catalyst/core/runner.py]: @asmekal yup, to make it easier to understand

Comment (reyzharkov) [reply catalyst/core/runner.py]: ok, renamed to `migrate_from_previous_stage`

**Pull request merged**